### PR TITLE
Cut v2.39.0-rc.0

### PR DESCRIPTION
**Pull request by codesome**
NOTE: this PR is open against main branch right now because there are a few PRs that I am still waiting to be merged. Once they are merged, I will create a release-2.39 branch and point this PR against that.

I am still waiting for the following things to be done

- [x] Temporarily remove `--enable-feature=metadata-storage` from this release since remote-write for metadata is not implemented yet. Flag to be added back with the remote-write PR. @tpaschalis has agreed to make that change.
- [x] Merge the following PRs
  - [x] #11296
  - [x] #11288
  - [x] #11340
  - [x] #11338
  - [x] #11331
  - [x] #11317

### CODE REVIEW

**Comment by codesome** [comment]
Making it ready for review. The CHANGELOG assumes that the PRs mentioned in the PR description have been merged (I will merge this PR after those are merged).

**Comment by beorn7** [comment]
What about #11317? It fixes a quite serious bug. We only need to review…

**Comment by codesome** [comment]
@rfratto does https://github.com/prometheus/prometheus/pull/9876 require a CHANGELOG entry here?

**Comment by codesome** [comment]
```
[BUGFIX] Agent: Fix validation of flag options and prevent WAL from growing more than desired. #9876
```

@rfratto could you confirm if the above entry looks fine? (from what I can see it is not just the default options as the PR title says)

**Comment by rfratto** [comment]
@codesome Looks good!

**Review by Nexucis — APPROVED** [review]
nice. Thank you @codesome 

As a side note, it would be interesting to sort the changelog entries per type. Like instead of

```
* [FEATURE] **experimental** TSDB: Add support for ingesting out-of-order samples. This is configured via `out_of_order_time_window` field in the config file; check config file docs for more info. #11075
* [ENHANCEMENT] API: `/-/healthy` and `/-/ready` API calls now also respond to a `HEAD` request on top of existing `GET` support. #11160 
* [ENHANCEMENT] PuppetDB SD: Add `__meta_puppetdb_query` label. #11238
* [ENHANCEMENT] AWS Lightsail SD: Add `__meta_ec2_region` label. #11326
* [ENHANCEMENT] AWS EC2 SD: Add `____meta_lightsail_region` label. #11326
* [ENHANCEMENT] TSDB: Improve WAL replay timings. #10973 #11307 #11319
* [ENHANCEMENT] Scrape: Optimise relabeling by re-using memory. #11147
* [ENHANCEMENT] TSDB: Optimise memory by not storing unnecessary data in the memory. #11280 #11288 #11296
* [ENHANCEMENT] UI: Click to copy label-value pair from query result to clipboard. #11229
* [ENHANCEMENT] TSDB: Allow overlapping blocks by default. `--storage.tsdb.allow-overlapping-blocks` now has no effect. #11331
* [BUGFIX] TSDB: Turn off isolation for Head compaction to fix a memory leak. #11317
* [BUGFIX] PromQL: Properly close file descriptor when logging unfinished queries. #11148
* [BUGFIX] TSDB: Fix 'invalid magic number 0' error on Prometheus startup. #11338
* [BUGFIX] Agent: Fix validation of flag options and prevent WAL from growing more than desired. #9876
```

we have 

```
* [FEATURE] **experimental** TSDB: Add support for ingesting out-of-order samples. This is configured via `out_of_order_time_window` field in the config file; check config file docs for more info. #11075
* [ENHANCEMENT] API: `/-/healthy` and `/-/ready` API calls now also respond to a `HEAD` request on top of existing `GET` support. #11160 
* [ENHANCEMENT] PuppetDB SD: Add `__meta_puppetdb_query` label. #11238
* [ENHANCEMENT] AWS Lightsail SD: Add `__meta_ec2_region` label. #11326
* [ENHANCEMENT] AWS EC2 SD: Add `____meta_lightsail_region` label. #11326
* [ENHANCEMENT] Scrape: Optimise relabeling by re-using memory. #11147
* [ENHANCEMENT] TSDB: Improve WAL replay timings. #10973 #11307 #11319
* [ENHANCEMENT] TSDB: Optimise memory by not storing unnecessary data in the memory. #11280 #11288 #11296
* [ENHANCEMENT] TSDB: Allow overlapping blocks by default. `--storage.tsdb.allow-overlapping-blocks` now has no effect. #11331
* [ENHANCEMENT] UI: Click to copy label-value pair from query result to clipboard. #11229
* [BUGFIX] TSDB: Turn off isolation for Head compaction to fix a memory leak. #11317
* [BUGFIX] TSDB: Fix 'invalid magic number 0' error on Prometheus startup. #11338
* [BUGFIX] PromQL: Properly close file descriptor when logging unfinished queries. #11148
* [BUGFIX] Agent: Fix validation of flag options and prevent WAL from growing more than desired. #9876
```

**Comment by codesome** [comment]
Thanks, makes sense

**Review by beorn7 — APPROVED** [review]

**Comment by codesome** [comment]
Benchmark analysis:

I can clearly see the impact of disabling isolation for head compaction - the isolation watermark different panel does not spike anymore.

In terms of CPU/Memory, seems similar, nothing conclusive. @bboreham is this unexpected?

There was one small difference: see [this](http://prombench.prometheus.io/grafana/d/7gmLoNDmz/prombench?orgId=1&var-pr-number=11344&from=1664528159441&to=1664530073365), this PR seems to have a few more compactions at this time than the earlier release. The number of compactions should not have changed. All earlier compactions match.

**Comment by bboreham** [comment]
> In terms of CPU/Memory, seems similar, nothing conclusive. @bboreham is this unexpected?

Memory: #11280 removed 8 bytes per series; #11288 removed 16; #11296 removed 56.
However #11075 added 48 bytes per series (assuming out-of-order function is not used).
So we should come out 32 bytes per series ahead.
The benchmark has around 5 million series, so we should save 150MB of heap, doubled to 300MB (by Go runtime allowing 100% growth).
The Next GC number is around 20GB, so we are looking for a 1.5% improvement.
I agree it's hard to see in the charts.

CPU I would have to look in more detail at profiles of PromBench running to say whether it is hitting any of the things that were improved.

**Comment by codesome** [comment]
Something that was not visible in the prombench https://twitter.com/_codesome/status/1577264183208398848

**Pull request merged**
